lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -776,56 +754,79 @@ def accessible_objects(user_or_attribute, operation, options = {})
     # ancestors of a specified root object or the object itself
     def accessible_ancestor_objects(user_or_attribute, operation, root_object, options = {})
       # If the root_object is a generic PM::Object, convert it the appropriate storage adapter Object
-      root_object_stored_pe = root_object.try(:stored_pe) || root_object
+      root_object = root_object.try(:stored_pe) || root_object
 
       # The final set of accessible objects must be ancestors of the root_object; avoid
       # duplicate ancestor calls when possible
       ancestor_objects = options[:ancestor_objects]
-      ancestor_objects ||= root_object_stored_pe.ancestors(type: class_for_type('object')) + [root_object_stored_pe]
+      ancestor_objects ||= root_object.ancestors(type: class_for_type('object')) + [root_object]
 
       # Short-circuit and return all ancestors (minus prohibitions) if the user_or_attribute
       # is authorized on the root node
-      if is_privilege?(user_or_attribute, operation, root_object_stored_pe)
-        return all_ancestor_objects(user_or_attribute, operation, root_object_stored_pe, ancestor_objects, options)
+      if is_privilege?(user_or_attribute, operation, root_object)
+        return all_ancestor_objects(user_or_attribute, operation, root_object, ancestor_objects, options)
       end
 
-      # Fetch all of the PEAs using the given UA or its descendants
-      user_attribute_ids = user_or_attribute.descendants.pluck(:id) | [user_or_attribute.id]
-      associations = PolicyElementAssociation.where(user_attribute_id: user_attribute_ids)
-      operation_set_ids = associations.pluck(:operation_set_id)
+      full_scope = objects_for_user_and_operation(user_or_attribute, operation, options)
+      candidates = full_scope & ancestor_objects
+
+      if options[:ignore_prohibitions] || !(prohibition = prohibition_for(operation))
+        candidates
+      else
+        preloaded_options = options.merge(ignore_prohibitions: true, ancestor_objects: ancestor_objects)
+        candidates - accessible_ancestor_objects(user_or_attribute, prohibition, root_object, preloaded_options)
+      end
+    end
 
-      # Narrow the list of PEAs to just those containing the specified operation
+    private
+
+    # Returns an array of all the objects accessible for a given user and operation
+    def objects_for_user_and_operation(user_or_attribute, operation, options)
+      associations = associations_for_user_or_attribute(user_or_attribute)
+      filtered_associations = associations_filtered_by_operation(associations, operation)
+      build_accessible_object_scope(filtered_associations, options)
+    end
+
+    # Gets the associations related to the given user or its descendants
```

Comment:
Users and user attributes are not the same thing. Our comments should reflect that.